onnxruntime/contrib_ops/rocm/diffusion/group_norm_tunable_op.h

```diff
@@ -0,0 +1,204 @@
+  break;
+
+  switch (params->cPerBlock) {
+    case 320:
+      LAUNCH_GROUPNORM_SCALE(256, 2)
+    case 480:
+      LAUNCH_GROUPNORM_SCALE(256, 2)
+    case 256:
+      LAUNCH_GROUPNORM_SCALE(128, 2)
+    case 128:
+      LAUNCH_GROUPNORM_SCALE(64, 2)
+    default:
+      ORT_NOT_IMPLEMENTED("Not implemented");
+  }
+}
+
+template <typename T, int ThreadsPerBlock, int VecSize>
+Status GroupNormNHWCScaleOp(const GroupNormNHWCParams<T>* params) {
+  dim3 grid;
+  grid.x = params->c / params->cPerBlock;
+  grid.y = CeilDiv(params->hw, params->hwPerBlock);
+  grid.z = params->n;
+
+  groupNormNHWCScaleKernel<T, ThreadsPerBlock, VecSize>
+      <<<grid, ThreadsPerBlock, 0, params->stream>>>(
+          params->dst, params->src, params->gamma, params->beta, params->redBuffer, params->epsilon, params->c, params->cPerBlock,
+          params->cPerGroup, params->groups, params->hwc, params->invHWC, params->hw, params->hwPerBlock, params->withSwish);
+  return HIP_CALL(hipGetLastError());
+}
+
+template <typename T, int ThreadsPerBlock, int VecSize>
+class GroupNormNHWCOp {
+ public:
+  Status operator()(const GroupNormNHWCParams<T>* params) {
+    DUMP_TENSOR_INIT();
+    DUMP_TENSOR("input", params->src, params->n, params->c, params->h * params->w);
+    DUMP_TENSOR("gamma", params->gamma, 1, params->c);
+    DUMP_TENSOR("beta", params->beta, 1, params->c);
+    HIP_RETURN_IF_ERROR(hipMemsetAsync(params->redBuffer, 0, GetGroupNormWorkspaceSizeInBytes(), params->stream));
+    auto status = GroupNormNHWCSumOp<T, ThreadsPerBlock, VecSize>(params);
+    ORT_RETURN_IF_ERROR(status);
+    DUMP_TENSOR("workspace", params->redBuffer, params->n, params->groups, 2);
+    HIP_RETURN_IF_ERROR(hipGetLastError());
+    status = GroupNormNHWCScaleOp<T, ThreadsPerBlock, VecSize>(params);
+    ORT_RETURN_IF_ERROR(status);
+    HIP_RETURN_IF_ERROR(hipGetLastError());
+    DUMP_TENSOR("output", params->dst, params->n, params->c, params->h * params->w);
+    return Status::OK();
+  }
+
+  Status IsSupported(const GroupNormNHWCParams<T>* params) {
+    TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(
+        !(params->c % VecSize == 0 && params->cPerGroup % VecSize == 0));
+    TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(!(params->cPerBlock % params->cPerGroup == 0 &&
+                                                params->c % params->cPerBlock == 0 &&
+                                                params->hw % params->hwPerBlock == 0));
+    TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(!(params->cPerBlock <= ThreadsPerBlock * VecSize &&
+                                                params->cPerBlock > (ThreadsPerBlock - GPU_WARP_SIZE) * VecSize));
+
+    return Status::OK();
+  }
+};
+
+template <typename T>
+Status GroupNormNHWCStaticSelection(const GroupNormNHWCParams<T>* params) {
+  DUMP_TENSOR_INIT();
+  DUMP_TENSOR("input", params->src, params->n, params->c, params->h * params->w);
```

Comment:
> DUMP_TENSOR

Same here: Is this leftover debugging code?

Reply:
These are some dump mechanism from CUDA, like CUDA's Attention. It is not mature. But I think it will be useful when you need to debug all thoes intermediate values. @PeixuanZuo You decide, if you think this kernel is complex enough, like Attention, then leave it, otherwise, clean it up would be a better choice.

Reply:
Thanks for your advice, I think dump values used here are not intermediate values, we can debug them using kernel_explorer. I will remove them.